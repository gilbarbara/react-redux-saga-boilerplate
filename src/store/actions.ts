export { alertHide, alertShow } from './slices/alerts';
export { setAppOptions } from './slices/app';
export { getRepos, getReposFailure, getReposSuccess } from './slices/github';
export { login, loginSuccess, logOut, logOutSuccess } from './slices/user';
