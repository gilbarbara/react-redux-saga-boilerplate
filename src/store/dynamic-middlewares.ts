import { compose, Middleware, MiddlewareAPI } from 'redux';

export const createDynamicMiddlewares = () => {
  let allDynamicMiddlewares: Middleware[] = [];
  let allAppliedDynamicMiddlewares: Middleware[] = [];
  let store: MiddlewareAPI;

  const enhancer: Middleware = _store => {
    store = _store;

    return next => action => {
      // @ts-ignore
      return compose(...allAppliedDynamicMiddlewares)(next)(action);
    };
  };

  const addMiddleware = (...middlewares: Middleware[]) => {
    // TODO: Fix this
    // @ts-ignore
    allAppliedDynamicMiddlewares.push(...middlewares.map(middleware => middleware(store)));
    allDynamicMiddlewares.push(...middlewares);
  };

  const removeMiddleware = (middleware: Middleware) => {
    const index = allDynamicMiddlewares.findIndex(d => d === middleware);

    if (index === -1) {
      // eslint-disable-next-line no-console
      console.error('Middleware does not exist!', middleware);

      return;
    }

    allDynamicMiddlewares = allDynamicMiddlewares.filter((_, mdwIndex) => mdwIndex !== index);
    allAppliedDynamicMiddlewares = allAppliedDynamicMiddlewares.filter(
      (_, mdwIndex) => mdwIndex !== index,
    );
  };

  const resetMiddlewares = () => {
    allAppliedDynamicMiddlewares = [];
    allDynamicMiddlewares = [];
  };

  return {
    enhancer,
    addMiddleware,
    removeMiddleware,
    resetMiddlewares,
  };
};

const dynamicMiddlewaresInstance = createDynamicMiddlewares();

export default dynamicMiddlewaresInstance.enhancer;

export const { addMiddleware, removeMiddleware, resetMiddlewares } = dynamicMiddlewaresInstance;
