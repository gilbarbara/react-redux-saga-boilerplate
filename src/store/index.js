import { applyMiddleware, createStore, compose, combineReducers } from 'redux';
import { persistStore, persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage';

import rootSaga from 'sagas/index';
import rootReducer from 'reducers/index';

import middleware, { sagaMiddleware } from './middleware';

const reducer = persistReducer(
  {
    key: 'rrsb', // key is required
    storage, // storage is now required
    whitelist: ['app', 'user'],
  },
  combineReducers({ ...rootReducer }),
);

const composeEnhancer = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

/* istanbul ignore next */
const configStore = (initialState = {}) => {
  const store = createStore(reducer, initialState, composeEnhancer(applyMiddleware(...middleware)));

  sagaMiddleware.run(rootSaga);

  if (module.hot) {
    module.hot.accept('reducers', () => {
      store.replaceReducer(require('reducers/index').default);
    });
  }

  return {
    persistor: persistStore(store),
    store,
  };
};

const { store, persistor } = configStore();

global.store = store;

export { store, persistor };
