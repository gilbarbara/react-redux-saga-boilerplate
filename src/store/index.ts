import { applyMiddleware, compose, createStore, Middleware, Store } from 'redux';
import { persistCombineReducers, persistStore } from 'redux-persist';
import autoMergeLevel2 from 'redux-persist/lib/stateReconciler/autoMergeLevel2';
import storage from 'redux-persist/lib/storage';

import reducers from 'reducers';
import rootSaga from 'sagas';

import { StoreState } from 'types';

import middleware, { sagaMiddleware } from './middleware';

const rootReducer = persistCombineReducers<StoreState>(
  {
    key: 'rrsb',
    stateReconciler: autoMergeLevel2,
    storage,
    whitelist: ['app', 'user'],
    timeout: 0,
  },
  reducers,
);

const composeEnhancer = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;

/* istanbul ignore next */
export const configStore = (initialState: any = {}, additionalMiddleware: Middleware[] = []) => {
  const store: Store = createStore(
    rootReducer,
    initialState,
    composeEnhancer(applyMiddleware(...additionalMiddleware, ...middleware)),
  );

  sagaMiddleware.run(rootSaga);

  if (module.hot) {
    module.hot.accept('reducers', () => {
      store.replaceReducer(rootReducer);
    });
  }

  return {
    persistor: persistStore(store),
    store,
  };
};
