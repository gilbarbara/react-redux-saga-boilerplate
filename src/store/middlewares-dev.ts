import { Middleware } from 'redux';
import invariant from 'redux-immutable-state-invariant';
import { createLogger } from 'redux-logger';

export default [invariant(), createLogger({ collapsed: true })] as Middleware[];
