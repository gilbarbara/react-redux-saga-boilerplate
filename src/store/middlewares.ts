import createSagaMiddleware from 'redux-saga';

export const sagaMiddleware = createSagaMiddleware();

const middlewares = [sagaMiddleware];

/* istanbul ignore next */
if (process.env.NODE_ENV === 'development') {
  const devMiddlewares = require('./middlewares-dev').default;

  middlewares.push(...devMiddlewares);
}

export default middlewares;
