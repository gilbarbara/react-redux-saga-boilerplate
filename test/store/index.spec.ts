import { configStore, store } from '~/store';

describe('store', () => {
  it('should create a new store', () => {
    expect(configStore().getState()).toMatchSnapshot();
  });

  it('should return the global store', () => {
    expect(store.getState()).toMatchSnapshot();
  });
});
