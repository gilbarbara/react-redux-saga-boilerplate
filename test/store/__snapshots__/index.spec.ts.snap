// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store should be able to create a new store 1`] = `
Object {
  "_persist": Object {
    "rehydrated": false,
    "version": -1,
  },
  "app": Object {
    "alerts": Array [],
  },
  "github": Object {
    "query": "",
    "topics": Object {},
  },
  "user": Object {
    "isAuthenticated": false,
    "status": "idle",
  },
}
`;
